driver-core/src/main/com/mongodb/internal/time/TimePoint.java

```diff
@@ -74,11 +78,40 @@ static TimePoint infinite() {
         return at(null);
     }
 
+    @Override
+    public Timeout shortenBy(final long amount, final TimeUnit timeUnit) {
+        if (nanos == null) {
+            return this; // shortening (lengthening) an infinite timeout does nothing
+        }
+        long durationNanos = NANOSECONDS.convert(amount, timeUnit);
+        return TimePoint.at(nanos - durationNanos);
+    }
+
+    @Override
+    public <T> T run(final TimeUnit timeUnit, final Supplier<T> onInfinite, final LongFunction<T> onHasRemaining,
+            final Supplier<T> onExpired) {
+        return checkedRun(timeUnit, onInfinite::get, onHasRemaining::apply, onExpired::get);
+    }
+
+    @Override
+    public <T, E extends Exception> T checkedRun(final TimeUnit timeUnit,
+            final CheckedSupplier<T, E> onInfinite, final CheckedFunction<Long, T, E> onHasRemaining,
+            final CheckedSupplier<T, E> onExpired) throws E {
+        if (this.isInfiniteLocal()) {
+            return onInfinite.get();
+        }
+        long remaining = remainingLocal(timeUnit);
+        if (remaining <= 0) {
+            return onExpired.get();
+        } else {
+            return onHasRemaining.apply(remaining);
+        }
+    }
+
     /**
      * @return true if this timepoint is infinite.
      */
-    @Override
-    public boolean isInfinite() {
+    public boolean isInfiniteLocal() {
```

Comment:
I refactored the implementing method, but kept the base method. The intent is to remove the base method, once we have consensus on the approach, and then this can be changed back.